### PR TITLE
Characters to leaves

### PR DESCRIPTION
**Pull request by zhujinxuan**
Feature: memory save by replacing characters with leaves in data structure
Currently, this PR is ready for review.

### CODE REVIEW

**Comment by zhujinxuan** [comment]
TODO: 
1. add tests for get-active-mark
2. add tests for merge leaves
3. add tests for insertLeaf
4. add tests for merge leaves
5. await #1808 
6. add method to splitText
7. rewrite Node.splitNode
8. search other characters
9. rewrite all text.js

**Comment by zhujinxuan** [comment]
```

  benchmarks
    changes
      delete-backward
        user : 413.96 → 428.08 ops/sec
        hr : 756.08 → 757.10 ops/sec
      delete-forward
        user : 133.94 → 144.41 ops/sec
        hr : 212.26 → 256.91 ops/sec
      insert-node-by-key
        user : 249.78 → 258.66 ops/sec
        hr : 456.35 → 471.37 ops/sec
      insert-text-by-key-multiple
        user : 73.47 → 83.89 ops/sec
        hr : 108.27 → 121.55 ops/sec
      insert-text-by-key
        user : 410.15 → 494.69 ops/sec
        hr : 725.70 → 795.48 ops/sec
      insert-text
        user : 1791.82 → 2258.47 ops/sec
        hr : 3223.50 → 4207.93 ops/sec
      normalize
        user : 765.91 → 704.66 ops/sec
        hr : 1225.53 → 1237.51 ops/sec
      remove-node-by-key
        user : 293.21 → 347.16 ops/sec
        hr : 526.84 → 595.46 ops/sec
      split-block
        user : 430.18 → 469.20 ops/sec
        hr : 730.98 → 763.30 ops/sec
    models
      from-json-big
        user : 4.15 → 5.54 ops/sec (33.31% faster)
        hr : 5.89 → 7.97 ops/sec (35.23% faster)
      from-json
        user : 169.57 → 327.19 ops/sec (92.96% faster)
        hr : 284.03 → 636.90 ops/sec (124.24% faster) 🙌
      get-active-marks-at-range
        user : 4603.62 → 6286.12 ops/sec (36.55% faster)
        hr : 6934.05 → 9327.45 ops/sec (34.52% faster)
      get-ancestors
        user : 5029.36 → 4694.56 ops/sec
        hr : 6960.38 → 7055.48 ops/sec
      get-blocks-at-range
        user : 681.36 → 798.74 ops/sec
        hr : 1278.29 → 1474.71 ops/sec
      get-blocks
        user : 6990.17 → 8562.95 ops/sec
        hr : 11638.49 → 12376.32 ops/sec
      get-characters-at-range
        user : 1292.18 → 320.15 ops/sec (304.00% slower) 😟
        hr : 1997.87 → 541.08 ops/sec (269.00% slower) 😟
      get-characters
        user : 3083.27 → 560.76 ops/sec (450.00% slower) 😟
        hr : 4255.75 → 770.84 ops/sec (452.00% slower) 😟
      get-common-ancestor
        user : 966.60 → 1086.86 ops/sec
        hr : 1338.08 → 1437.77 ops/sec
      get-furthest-ancestor
        user : 7548.35 → 8605.46 ops/sec
        hr : 12516.56 → 13014.25 ops/sec
      get-inlines-at-range
        user : 665.99 → 713.12 ops/sec
        hr : 804.96 → 898.70 ops/sec
      get-inlines
        user : 10365.85 → 13559.89 ops/sec
        hr : 15570.90 → 19191.86 ops/sec
      get-leaves
        user : 7848.32 → 776260.41 ops/sec (9790.79% faster) 😱
        hr : 14465.28 → 1279971.74 ops/sec (8748.58% faster) 😱
      get-marks-at-range
        user : 616.05 → 414.85 ops/sec (48.00% slower)
        hr : 1082.59 → 837.13 ops/sec
      get-marks
        user : 1866.85 → 2292.86 ops/sec
        hr : 2670.76 → 3703.34 ops/sec (38.66% faster)
      get-parent
        user : 7024.27 → 6496.41 ops/sec
        hr : 8331.50 → 8276.03 ops/sec
      get-path
        user : 2366.90 → 2343.16 ops/sec
        hr : 3369.26 → 3333.90 ops/sec
      get-texts-at-range
        user : 3508.45 → 5416.59 ops/sec (54.39% faster)
        hr : 5788.52 → 6779.09 ops/sec
      get-texts
        user : 11208.40 → 14869.31 ops/sec
        hr : 16970.83 → 20040.89 ops/sec
      getDescendant
        user : 9176.62 → 8883.55 ops/sec
        hr : 13355.21 → 13025.06 ops/sec
      has-node-multiple
        user : 2055.68 → 2173.20 ops/sec
        hr : 2639.94 → 2615.64 ops/sec
      has-node
        user : 9077.04 → 10249.72 ops/sec
        hr : 13249.64 → 13718.36 ops/sec
      to-json
        user : 726.83 → 2343.29 ops/sec (222.40% faster) 🙌
        hr : 1192.27 → 3542.70 ops/sec (197.14% faster) 🙌
      update-node
        user : 3177.19 → 3497.95 ops/sec
        hr : 4893.11 → 4978.00 ops/sec
    html-serializer
      deserialize
        user : 8.96 → 6.59 ops/sec (36.00% slower)
        hr : 12.29 → 8.66 ops/sec (42.00% slower)
      serialize
        user : 236.38 → 469.60 ops/sec (98.66% faster)
        hr : 386.76 → 792.09 ops/sec (104.80% faster) 🙌
    plain-serializer
      deserialize
        user : 457.53 → 2286.87 ops/sec (399.83% faster) 🙌
        hr : 822.85 → 4052.55 ops/sec (392.50% faster) 🙌
      serialize
        user : 2112.30 → 3865.40 ops/sec (82.99% faster)
        hr : 3538.24 → 5973.32 ops/sec (68.82% faster)
    rendering
      normal
        user : 71.61 → 78.17 ops/sec
        hr : 104.96 → 114.36 ops/sec
```

**Comment by Soreine** [comment]
This is really interesting :)

**Comment by zhujinxuan** [comment]
@Soreine The rendering is not speed up significantly, which is strange to me.  We may need some profiling about what the bottle-neck of the rendering. (Perhaps Stack logic?)

**Comment by zhujinxuan** [comment]
in recent pr, delete backward speed up:
```
      delete-backward
        User Space CPU Time Measure: : 357.38 → 557.67 ops/sec (56.04% faster)
```

**Comment by zhujinxuan** [comment]
@ianstormtaylor This PR is ready for review~

File: packages/slate/src/models/node.js [hunk 1]
```diff
@@ -1944,7 +1944,7 @@ class Node {
       throw new Error(`Could not find a descendant node with key "${key}".`)
 
     const index = parent.nodes.findIndex(n => n.key === key)
-    const nodes = parent.nodes.splice(index, 1)
+    const nodes = parent.nodes.delete(index)
```
Comment (ianstormtaylor) [inline packages/slate/src/models/node.js]: Nice simplifications here!

File: packages/slate/src/models/node.js [hunk 2]
```diff
@@ -1980,10 +1980,7 @@ class Node {
     // If the child is a text node, the `position` refers to the text offset at
     // which to split it.
     if (child.object == 'text') {
-      const befores = child.characters.take(position)
-      const afters = child.characters.skip(position)
-      one = child.set('characters', befores)
-      two = child.set('characters', afters).regenerateKey()
+      ;[one, two] = child.splitText(position)
```
Comment (ianstormtaylor) [inline packages/slate/src/models/node.js]: Is this `;` necessary?

File: packages/slate/src/models/text.js [hunk 3]
```diff
@@ -16,9 +14,10 @@ import memoize from '../utils/memoize'
  */
 
 const DEFAULTS = {
-  characters: new List(),
+  leaves: List(),
   key: undefined,
 }
+const emptyLeaf = Leaf.create({})
```
Comment (ianstormtaylor) [inline packages/slate/src/models/text.js]: What's this doing here? Why is it not in `CONSTANT_CASE`? Why does it not have a docblock comment? Why is there no line break before it?

File: packages/slate/src/models/text.js [hunk 4]
```diff
@@ -164,7 +165,53 @@ class Text extends Record(DEFAULTS) {
...
+   * Get the concatenated text of the node, cached for text getter
+   *
+   * @returns {String}
+   */
+
+  getString() {
+    return this.leaves.reduce((string, leaf) => string + leaf.text, '')
+  }
+
+  get characters() {
```
Comment (ianstormtaylor) [inline packages/slate/src/models/text.js]: Docblock...

File: packages/slate/src/models/text.js [hunk 5]
```diff
@@ -241,82 +273,79 @@ class Text extends Record(DEFAULTS) {
...
-    }
-
-    // PERF: convert the leaves to immutable objects after iterating.
-    leaves = new List(leaves.map(object => new Leaf(object)))
-
-    // Return the leaves.
-    return leaves
+      }
+      leaves = leaves.map(x => x.addMarks(marks))
+      return
```
Comment (ianstormtaylor) [inline packages/slate/src/models/text.js]: Unnecessary?

File: packages/slate/test/models/text/delete/all-text-length/marked-text.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/** @jsx h */
+
+import h from '../../../../helpers/h'
+
+export const input = <b>Cat is Cute</b>[0]
+export default function(t) {
+  return t.removeText(0, t.text.length)
+}
+
+export const output = <b />[0]
```
Comment (ianstormtaylor) [inline packages/slate/test/models/text/delete/all-text-length/marked-text.js]: I'd like to revert this behavior. It should be the same as deleting all the characters above with italic and bold, where the end marks are none.

File: packages/slate/test/models/text/delete/all-text-length/partial-marked-text.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+export const input = (
+  <b>
+    Cat is <i>Cute</i>
+  </b>
+)[0]
+export default function(t) {
+  return t.removeText(0, t.text.length)
+}
+
+export const output = <b />[0]
```
Comment (ianstormtaylor) [inline packages/slate/test/models/text/delete/all-text-length/partial-marked-text.js]: Same here, deleting all the text should remove marks.

File: packages/slate/test/models/text/insert/from-start/pure-text-on-null-text-at-invalid-offset.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+/** @jsx h */
+
+import h from '../../../../helpers/h'
+
+export const input = <text />[0]
+export default function(t) {
```
Comment (ianstormtaylor) [inline packages/slate/test/models/text/insert/from-start/pure-text-on-null-text-at-invalid-offset.js]: Please put line breaks before all of these exports.

File: packages/slate/test/models/text/marks/get-marks/marked-text.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+import { Mark } from '../../../../../'
+
+export const input = (
+  <text>
+    <i>Cat is </i>
+    <b>Cute</b>
+  </text>
+)[0]
+
+export default function(t) {
```
Comment (ianstormtaylor) [inline packages/slate/test/models/text/marks/get-marks/marked-text.js]: This linebreak before this export is good. All the other test files should be like this.

File: packages/slate/src/models/leaf.js [hunk 10]
```diff
@@ -138,6 +235,10 @@ class Leaf extends Record(DEFAULTS) {
    */
 
   getCharacters() {
+    logger.deprecate(
+      'slate@0.35.0',
+      'The `characters` property of Slate objects has been removed.'
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/src/models/leaf.js]: Fixed~

File: packages/slate/src/models/leaf.js [hunk 11]
```diff
@@ -151,6 +252,48 @@ class Leaf extends Record(DEFAULTS) {
...
+   *
+   * @param {Mark} mark
+   * @param {Mark} newMark
+   * @returns {Leaf}
+   */
+
+  updateMark(mark, newMark) {
+    const { marks } = this
+    if (newMark.equals(mark)) return this
+    if (!marks || !marks.has(mark)) return this
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/src/models/leaf.js]: Fixed~ Thank you.

File: packages/slate/src/models/node.js (hunk 2 above)
Comment (zhujinxuan) [reply packages/slate/src/models/node.js]: Fixed; Thank you

Comment (zhujinxuan) [reply packages/slate/src/models/node.js]: Not fixed; prettier eslint suggest a `;` here.

File: packages/slate/src/models/text.js (hunk 3 above)
Comment (zhujinxuan) [reply packages/slate/src/models/text.js]: Fixed~ It is a pre-created Leaf to speed up empty text Leaf creation.

File: packages/slate/src/models/text.js (hunk 4 above)
Comment (zhujinxuan) [reply packages/slate/src/models/text.js]: Fixed~ Thank you.

File: packages/slate/src/models/text.js (hunk 5 above)
Comment (zhujinxuan) [reply packages/slate/src/models/text.js]: Yes, the return is unnecessary. Fixed~ Thank you.

File: packages/slate/test/models/text/delete/all-text-length/marked-text.js (hunk 6 above)
Comment (zhujinxuan) [reply packages/slate/test/models/text/delete/all-text-length/marked-text.js]: I am considering this case: 

If we have some blocks are marked with bold or other marks, like:
```
<b>
<p>First Line</p>
<p>Second Line</p>
<p>Third Line</p>
</b>
```
If someone delete the all content in the second line, then insert, under the revert, the user will get
```
<p><b>First Line</b></p>
<p>New Second Line</p>
<p><b>Third Line</b></p>
```

By keeping the mark, the user will get the second line bold.
```
<p><b>First Line</b></p>
<p><b>New Second Line</b></p>
<p><b>Third Line</b></p>
``` 
Some aged users tends to select all document, mark with times-new-roman font, and then starts editing.  They will think that, if a range is marked, then any insert in that range is always marked.  So I implement this feature for empty text.


The only potential bug of the empty text with mark is about merging text node and insert text; I have made the special treatment that, if the current leaf text is empty, then the mark will have no influence if the merged node or the insert text is not empty.

**Comment by zhujinxuan** [comment]
@ianstormtaylor  A question: consider the following code:

```
    if (Array.isArray(leaves)) {
      leaves = List(leaves.map(x => Leaf.create(x)))
    } else if (List.isList(leaves)) {
      leaves = leaves.map(x => Leaf.create(x))
    } else {
      throw new Error('leaves must be either List or Array')
    }
```
shall we use `throw new TypeError` here?

File: packages/slate/src/models/leaf.js [hunk 12]
```diff
@@ -49,6 +49,99 @@ class Leaf extends Record(DEFAULTS) {
...
+    })
+    if (!left) return [leaves, List()]
+
+    if (left.text === '') {
+      if (index === 0) {
+        return [List.of(left), leaves]
+      }
+      return [leaves.take(index), leaves.skip(index)]
+    }
+    if (right.text === '') {
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/src/models/leaf.js]: Fixed in leaf.js and node.js.  I am fixing files in `test/` right now.

**Comment by zhujinxuan** [comment]
line breaks fixed~

**Comment by zhujinxuan** [comment]
@ianstormtaylor Is there any updates I shall do?

**Comment by zhujinxuan** [comment]
@Soreine  Though it is un-necessary, can you add a `performance` tag to this PR?

**Comment by zhujinxuan** [comment]
@SamyPesse Hi, I have deleted the `get-characters-at-range` in https://github.com/ianstormtaylor/slate/pull/1765;  All benchmark things goes to PR1765 because current benchmark has bugs.

File: packages/slate/test/models/text/delete/all-text-length/marked-text.js (hunk 6 above)
Comment (ianstormtaylor) [reply packages/slate/test/models/text/delete/all-text-length/marked-text.js]: Good argument, it looks like Google Docs follows your logic here too. I'm convinced!

File: packages/slate/src/models/leaf.js [hunk 13]
```diff
@@ -99,7 +208,13 @@ class Leaf extends Record(DEFAULTS) {
...
 
   static isLeaf(any) {
-    return !!(any && any[MODEL_TYPES.LEAF])
+    if (!any || !any.__proto__) return false
+
+    if (any.__proto__.hasOwnProperty) {
+      return any.__proto__.hasOwnProperty(MODEL_TYPES.LEAF)
+    }
+
+    return !!any[MODEL_TYPES.LEAF]
```
Comment (ianstormtaylor) [inline packages/slate/src/models/leaf.js]: Why is this more complex like this?

File: packages/slate/src/models/leaf.js [hunk 14]
```diff
@@ -189,3 +350,9 @@ Leaf.prototype[MODEL_TYPES.LEAF] = true
  */
 
 export default Leaf
+
+/*
+ * An pre-created Leaf to speed up empty Leaf creation in some cases
+ */
+
+export const EMPTY_LEAF = Leaf.create({})
```
Comment (ianstormtaylor) [inline packages/slate/src/models/leaf.js]: Can you omit this an add it in a new pull request later? I don't want to add unnecessary things here. This should be proven by a separate pull request that it actually impacts the benchmarks.

File: packages/slate/src/models/node.js [hunk 15]
```diff
@@ -1166,26 +1164,29 @@ class Node {
...
+      const currentMark =
+        text.text === '' ? text.getMarksAtIndex(0) : text.getMarksAtIndex(1)
+      return currentMark.intersect(previousMarks)
     }
-
-    const text = this.getDescendant(key)
-    const char = text.characters.get(offset - 1)
-    if (!char) return OrderedSet()
-    return new OrderedSet(char.marks)
+    return text.getMarksAtIndex(offset)
```
Comment (ianstormtaylor) [inline packages/slate/src/models/node.js]: Please clean up this function.

**Comment by zhujinxuan** [comment]
@ianstormtaylor Hi, the two problems are fixed.

**Comment by ianstormtaylor** [comment]
Thank you @zhujinxuan, this is really really great work!

**Comment by Soreine** [comment]
Just want to say thank you @zhujinxuan once more. All this is great 🏄‍♂️

**Pull request merged**